site/en/docs/extensions/mv3/nativeMessaging/index.md

```diff
@@ -0,0 +1,202 @@
+0 if the calling context is a background script page.
+
+## Connecting to a native application {: #native-messaging-client }
+
+Sending and receiving messages to and from a native application is very similar to cross-extension
+messaging. The main difference is that [runtime.connectNative][connect-native] is used instead of
+[runtime.connect][connect], and [runtime.sendNativeMessage][send-native-message] is used instead of
+[runtime.sendMessage][send-message].
+These methods can only be used if the "nativeMessaging" permission is [declared][declare-permissions] in your extensions's
+manifest file.
+
+The following example creates a [runtime.Port][port] object that's connected to native messaging host
+`com.my_company.my_application`, starts listening for messages from that port and sends one outgoing
+message:
+
+```js
+var port = chrome.runtime.connectNative('com.my_company.my_application');
+port.onMessage.addListener(function (msg) {
+  console.log('Received' + msg);
+});
+port.onDisconnect.addListener(function () {
+  console.log('Disconnected');
+});
+port.postMessage({text: 'Hello, my_application'});
+```
+
+[runtime.sendNativeMessage][send-native-message] can be used to send a message to native application without creating
+a port, e.g.:
+
+```js
+chrome.runtime.sendNativeMessage(
+  'com.my_company.my_application',
+  {text: 'Hello'},
+  function (response) {
+    console.log('Received ' + response);
+  }
+);
+```
+
+## Debugging native messaging {: #native-messaging-debugging }
+
+When the native messaging host fails to start, writes to `stderr` or when it violates the
+communication protocol, output is written to the error log of Chrome. On Linux and macOS, this log
+can easily be accessed by starting Chrome from the command line and watching its output in the
+terminal. On Windows, use `--enable-logging` as explained at [How to enable logging][enable-logging].
+
+Here are some errors and tips for solving the issues:
+
+- Failed to start native messaging host.
+  - Check whether you have sufficient permissions to execute the file.
+- Invalid native messaging host name specified.
+  - Check whether the name contains any invalid characters. Only lowercase alphanumeric characters,
+    underscores and dots are allowed. A name cannot start or end with a dot, and a dot cannot be
+    followed by another dot.
+- Native host has exited.
+  - The pipe to the native messaging host was broken before the message was read by Chrome. This is most
+    likely initiated from your native messaging host.
+- Specified native messaging host not found.
+  - Is the name spelled correctly in the extension and in the manifest file?
+  - Is the manifest put in the right directory and with the correct name? See [native messaging host
+    location][native-messaging-host-location] for the expected formats.
+  - Is the manifest file in the correct format? In particular, is the JSON syntax correct and do the
```

Comment:
```suggestion
  - Is the manifest file in the correct format? In particular, is the JSON valid and well-formed and do the
```

Reply:
I actually think this makes it more wordy and hard to follow.

Reply:
Isn't 'valid and well-formed' the more technically correct term though?